tests/pdb_test.go

```diff
@@ -0,0 +1,93 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2024 The KubeVirt Contributors
+ *
+ */
+
+package tests_test
+
+import (
+	"context"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	expect "github.com/google/goexpect"
+	policyv1 "k8s.io/api/policy/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	v1 "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/pkg/libvmi"
+	"kubevirt.io/kubevirt/tests/console"
+	"kubevirt.io/kubevirt/tests/decorators"
+	"kubevirt.io/kubevirt/tests/framework/kubevirt"
+	"kubevirt.io/kubevirt/tests/framework/matcher"
+	"kubevirt.io/kubevirt/tests/libvmifact"
+	"kubevirt.io/kubevirt/tests/libvmops"
+	"kubevirt.io/kubevirt/tests/testsuite"
+)
+
+var _ = Describe("[sig-compute]Pod Disruption Budget (PDB)", decorators.SigCompute, func() {
+
+	It("should ensure the Pod Disruption Budget (PDB) is deleted when the VM is stopped via guest OS", func() {
+		var err error
+
+		By("Creating test VM")
+		vm := libvmi.NewVirtualMachine(
+			libvmifact.NewCirros(
+				libvmi.WithEvictionStrategy(v1.EvictionStrategyLiveMigrate),
+				libvmi.WithResourceMemory("128Mi"),
+			),
+			libvmi.WithRunStrategy(v1.RunStrategyManual),
+		)
+
+		vm, err = kubevirt.Client().VirtualMachine(testsuite.NamespaceTestDefault).Create(context.Background(), vm, metav1.CreateOptions{})
+		Expect(err).ToNot(HaveOccurred())
+
+		libvmops.StartVirtualMachine(vm)
+
+		By("Waiting for the VM to be ready")
+		Eventually(func() (*v1.VirtualMachineInstance, error) {
+			return kubevirt.Client().VirtualMachineInstance(testsuite.NamespaceTestDefault).Get(context.Background(), vm.Name, metav1.GetOptions{})
+		}, 180*time.Second, 1*time.Second).Should(BeAssignableToTypeOf(&v1.VirtualMachineInstance{}))
+
+		By("Verifying PDB for testvm existence")
+		Eventually(func() ([]policyv1.PodDisruptionBudget, error) {
+			pdbs, err := kubevirt.Client().PolicyV1().PodDisruptionBudgets(testsuite.NamespaceTestDefault).List(context.Background(), metav1.ListOptions{})
+			return pdbs.Items, err
+		}, 60*time.Second, 1*time.Second).Should(Not(BeEmpty()), "The Pod Disruption Budget should be created")
+
+		By("Stopping the VM from the guest OS")
+		vmi, err := kubevirt.Client().VirtualMachineInstance(testsuite.NamespaceTestDefault).Get(context.Background(), vm.Name, metav1.GetOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		Expect(console.LoginToCirros(vmi)).To(Succeed())
+		Expect(console.SafeExpectBatch(vmi, []expect.Batcher{
+			&expect.BSnd{S: "sudo poweroff\n"},
+			&expect.BExp{R: "The system is going down NOW!"},
```

Comment:
Won't this be racy?

Reply:
- Simplified error handling for Batcher calls to avoid test failures if the VMI shuts down fast.
- If the VMI is already in a final state, the test won't fail.

Reply:
@dasionov I also think it can be racy, to be on the safe side, can we do the following instead:
```
					By("Issuing a poweroff command from inside VM")
					Expect(console.SafeExpectBatch(vmi, []expect.Batcher{
						&expect.BSnd{S: "sudo poweroff\n"},
						&expect.BExp{R: console.PromptExpression},
					}, 10)).To(Succeed())

					By("Ensuring the VirtualMachineInstance enters Succeeded phase")
					Eventually(ThisVMI(vmi), 240*time.Second, 1*time.Second).Should(HaveSucceeded())
```